File: invert-binary-tree/jaejeong1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  // 풀이: child node로 내려가면서 재귀로 계속 left와 right를 바꿔준다.
  // TC: O(N)
  // SC: O(N)
  public TreeNode invertTree(TreeNode root) {
    return invert(root);
  }

  private TreeNode invert(TreeNode node) {
    if (node == null) {
      return node;
    }

    node = swap(node);

    invert(node.left);
    invert(node.right);

    return node;
  }

  private TreeNode swap(TreeNode node) {
    var temp = node.left;
    node.left = node.right;
    node.right = temp;

    return node;
  }
}

File: invert-binary-tree/jaejeong1.kt
class TreeNode(var `val`: Int) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}

class InvertBinaryTree {
    fun invertTree(root: TreeNode?): TreeNode? {
        // 이진 트리의 root 를 주면, 트리를 뒤집고, 그것의 Root를 반환해라
        // 재귀적으로 left, right 를 바꾸면 된다. child가 둘다 null 이면 종료
        // 시간복잡도: O(N), 공간복잡도: O(1)
        if (root == null) return null

        if (root.left == null && root.right == null) {
            return root
        }

        val temp = root.left
        root.left = root.right
        root.right = temp

        invertTree(root.left)
        invertTree(root.right)

        return root
    }

}
